test/jdk/java/lang/ClassLoader/loadLibraryUnload/libloadLibraryUnload.c

```diff
@@ -40,9 +42,25 @@ JNI_OnLoad(JavaVM *vm, void *reserved)
     return JNI_VERSION_1_2;
 }
 
+JNIEXPORT void JNICALL Java_p_Class1_setRef(JNIEnv *env, jobject this, jobject obj) {
+    if (ref == NULL) {
+        // Only create a single GlobalRef
+        ref = (*env)->NewGlobalRef(env, obj);
```

Comment:
Hello Roger, do we not use any synchronization here because even if both winning threads end up concurrently here, the `Object` instance they will be setting as a global ref will be the same one? Ofcourse, there will then be two calls to `NewGlobalRef` for the same instance. I am unaware what the semantics of that call will be in such cases and if it is of any concern.

Reply:
Its a racy bug, each NewGlobalRef will have a reference to the object and only the global ref in the last call to `setRef` will be cleared.
The easiest fix is to synchronize on `Class1.class` in Class1.loadLibrary().